pepsico/monthly_seasonal_conv_mapping.py

```diff
@@ -0,0 +1,111 @@
+from urllib.request import urlopen
+import json
+
+def compute_annual_monthly_avg(ds, variable, start_year=None, end_year=None):
+    #compute the avg (variable) for each month across the specified years
+    
+    #if precipitation variable, change from kg to mm per day
+    if variable == 'pr':
+        ds[variable] *= 86400
+    
+    if start_year and end_year:
+        ds = ds.sel(T=slice(f"{start_year}-01-01", f"{end_year}-12-31"))
+
+    # Convert daily data to monthly
+    monthly_ds = ds.resample(T="1M").mean()
+
+    # Compute the average for each month across all years
+    monthly_avg = monthly_ds.mean(dim=["X", "Y"])
+    annual_monthly_avg = monthly_avg.groupby("T.month").mean(dim="T")
+    df = annual_monthly_avg.to_dataframe().reset_index()
+    return df
+
+def compute_annual_seasonal_avg(monthly_ds, variable, start_year=None, end_year=None):
+    #compute monthly avgs across all years
+    monthly_avg = monthly_ds.mean(dim=["X", "Y"])
+
+    # Compute rolling seasonal average (3-month rolling window, center-aligned)
+    rolling_avg = monthly_avg.rolling(T=3, center=True).mean()
+    
+    # Compute the average variable for each rolling season across all years
+    rolling_seasonal_avg = rolling_avg.groupby("T.month").mean(dim="T")
+    
+    # Create a DataFrame for seasonal averages
+    seasonal_avgs_df = rolling_seasonal_avg.to_dataframe().reset_index()
+
+    # Assign the season column with correct length
+    seasonal_avgs_df['season'] = ['DJF', 'JFM', 'FMA', 'MAM', 'AMJ', 'MJJ',
+                                   'JJA', 'JAS', 'ASO', 'SON', 'OND', 'NDJ']
+
+    return seasonal_avgs_df
+    
+
+def write_to_csv(df, scenario, model, variable, output_dir):
+
+    # Write to CSV file
+    file_path = f'{output_dir}/{scenario}_{model}_{variable}_annual_monthly_avg.csv'
+    df.to_csv(file_path, index=False)
+
+
+def map_averages(df, model, variable):
+
+    df = df.sort_values(by='month')
+    map_fig = px.imshow(
+        df.pivot(index='Y', columns='X', values=variable), 
+        labels={'color': 'Kelvins'},
+        title=f'{variable} Monthly Averages',
+        origin='lower'  # Set the origin to lower to flip the y-axis if needed
+    )
+    return fig
+
+    
+
+def main(scenario, model, variable, start_year=None, end_year=None, output_dir='/home/sz3116/python-maprooms/pepsico/resources'):
+    #main to run functions
+
+    # Open zarr file 
+    ds = xr.open_zarr(f'//Data/data24/ISIMIP3b/InputData/climate/atmosphere/bias-adjusted/global/daily/historical/{model}/zarr/{variable}')
+
+    print("\nVariable attributes:")
+    for var in ds.data_vars:
+        print(f"\nAttributes for variable '{var}':")
+        print(ds[var].attrs)
+    
+    # Compute the average variable for each month across all years
+    annual_monthly_avg = compute_annual_monthly_avg(ds, variable, start_year, end_year)
+    
+    #call seasonal function
+    rolling_seasonal_avg = compute_annual_seasonal_avg(ds, variable, start_year, end_year)
```

Comment:
compute_annual_seasonal_avg takes a monthly ds. At this point, ds is still the daily data. Maybe what main should do is:
1/ read data according to scenario, model, variable
2/ select data in years
3/ compute monthly averages (have the output be xarray and not df -- the df conversion is really only for writing and plotting)
4/ compute seasonal ones from monthly ones (have the output be xarray and not df -- the df conversion is really only for writing and plotting)
5/ average in space, convert to df and write to csv both outputs of 3 and 4 on one hand
6/ select a month or season from 3 or 4, convert to df and map

I am not entirely sure what end result(s) you are aiming at so the above might not be quite it, but closer.

Reply:
I have pushed the adjusted code based on these comments. Not sure if it is exactly what you were asking. I broke the code down into more functions so it was more xarray outputs instead of df. Also, I have monthly annual and seasonal annual maps based on a specific month a user inputs in the main method.